cob_twist_controller/include/cob_twist_controller/constraints/constraint_impl.h

```diff
@@ -211,98 +549,441 @@ double JointLimitAvoidance<T_PARAMS, PRIO>::getActivationThreshold() const
+    const double limit_min = std::abs(params.limits_min[joint_idx]);
+    const double limit_max = std::abs(params.limits_max[joint_idx]);
+    const double joint_pos = this->joint_states_.current_q_(joint_idx);
+
+    this->abs_delta_max_ = std::abs(limit_max - joint_pos);
+    this->rel_max_ = std::abs(this->abs_delta_max_ / limit_max);
+
+    this->abs_delta_min_ = std::abs(joint_pos - limit_min);
+    this->rel_min_ = std::abs(this->abs_delta_min_ / limit_min);
+
+    this->calcValue();
+    this->calcPartialValues();
+    this->calcDerivativeValue();
+
+    double activation = this->getActivationThreshold();
+    double critical = 0.5 * activation;
+
+    const double rel_val = this->rel_max_ < this->rel_min_ ? this->rel_max_ : this->rel_min_;
+
+    if(rel_val < critical)
+    {
+        this->state_.setState(CRITICAL); // -> avoid HW destruction.
+    }
+    else if(rel_val < activation)
+    {
+        this->state_.setState(DANGER); // GPM
+    }
+    else
+    {
+        this->state_.setState(NORMAL);
+    }
+}
+
+/// Calculate values of the JLA cost function.
+template <typename T_PARAMS, typename PRIO>
+double JointLimitAvoidanceIneq<T_PARAMS, PRIO>::calcValue()
+{
+    const TwistControllerParams& params = this->constraint_params_.getParams();
+    int32_t joint_idx = this->constraint_params_.joint_idx_;
+    double limit_min = params.limits_min[joint_idx];
+    double limit_max = params.limits_max[joint_idx];
+
+    double joint_pos = this->joint_states_.current_q_(joint_idx);
+
+    this->last_value_ = this->value_;
+    this->value_ = (limit_max - joint_pos) * (joint_pos - limit_min);
+    return this->value_;
+}
+
+/**
+ * Simple first order differential equation for exponential increase (move away from limit!)
+ * @return 1st order diff equation value.
+ */
+template <typename T_PARAMS, typename PRIO>
+double JointLimitAvoidanceIneq<T_PARAMS, PRIO>::calcDerivativeValue()
+{
+    this->derivative_value_ = 0.1 * this->value_;
+    return this->derivative_value_;
+}
+
+/// Calculates values of the gradient of the cost function
+template <typename T_PARAMS, typename PRIO>
+Eigen::VectorXd JointLimitAvoidanceIneq<T_PARAMS, PRIO>::calcPartialValues()
+{
+    const TwistControllerParams& params = this->constraint_params_.getParams();
+    int32_t joint_idx = this->constraint_params_.joint_idx_;
+    double limit_min = params.limits_min[joint_idx];
+    double limit_max = params.limits_max[joint_idx];
+    double joint_pos = this->joint_states_.current_q_(joint_idx);
+    Eigen::VectorXd partial_values = Eigen::VectorXd::Zero(this->jacobian_data_.cols());
+    partial_values(this->constraint_params_.joint_idx_) = -2.0 * joint_pos + limit_max + limit_min;
+    this->partial_values_ = partial_values;
+    return this->partial_values_;
+}
+
+
+/// Returns the threshold of the cost function to become active.
+template <typename T_PARAMS, typename PRIO>
+double JointLimitAvoidanceIneq<T_PARAMS, PRIO>::getActivationThreshold() const
```

Comment:
As it seems that it occurs at several occasions:
should `activation_threshold` be a (dynamic) configurable parameter? or is it just different constraint types but fixed within those? 
To be changed wherever it applies!